File: MarkdownEditor/MarkdownEditor/Sources/MainWindowController.m
//
//  MainWindowController.m
//  MarkdownEditor
//
//  Created by Iwaki Satoshi on 2018/03/03.
//  Copyright © 2018 Satoshi Iwaki and 2026 Vincent Janelle. All rights reserved.
//

#import "MainWindowController.h"

@interface MainWindowController () {
}

@end

@implementation MainWindowController

static const NSSize MainWindowMinimumContentSize = {1200.0, 800.0};

- (NSArray<NSString *> *)converters {
    return ConverterManager.sharedInstance.converters;
}

- (NSUInteger)selectedConverterIndex {
    return ConverterManager.sharedInstance.selectedConverterIndex;
}

- (void)setSelectedConverterIndex:(NSUInteger)selectedFormatIndex {
    ConverterManager.sharedInstance.selectedConverterIndex = selectedFormatIndex;
}

- (void)windowDidLoad {
    [super windowDidLoad];
    self.window.contentMinSize = MainWindowMinimumContentSize;
    if (self.window.contentView.frame.size.width < MainWindowMinimumContentSize.width ||
        self.window.contentView.frame.size.height < MainWindowMinimumContentSize.height) {
        [self.window setContentSize:MainWindowMinimumContentSize];
        [self.window center];
    }
}

@end

File: MarkdownEditor/MarkdownEditor/Sources/MainWindowController.swift
import Cocoa

@objc(MainWindowController)
class MainWindowController: NSWindowController, NSWindowDelegate {
    private static let minimumContentSize = NSSize(width: 720.0, height: 480.0)
    private static let frameAutosaveName = "MarkdownEditorMainWindow"
    private var confirmedCloseInProgress = false

    @objc var converters: [String] {
        ConverterManager.shared.converters
    }

    @objc var selectedConverterIndex: UInt {
        get { UInt(ConverterManager.shared.selectedConverterIndex) }
        set { ConverterManager.shared.selectedConverterIndex = Int(newValue) }
    }

    override func windowDidLoad() {
        super.windowDidLoad()
        windowFrameAutosaveName = Self.frameAutosaveName
        window?.contentMinSize = Self.minimumContentSize
        guard let window else {
            return
        }
        let contentSize = window.contentView?.frame.size ?? .zero
        if contentSize.width < Self.minimumContentSize.width ||
            contentSize.height < Self.minimumContentSize.height {
            window.setContentSize(Self.minimumContentSize)
            window.center()
        }
    }

    func windowShouldClose(_ sender: NSWindow) -> Bool {
        guard !confirmedCloseInProgress,
              let editorViewController = findEditorViewController(in: contentViewController),
              editorViewController.dirty else {
            return true
        }

        editorViewController.confirmClosing { [weak self, weak sender] shouldClose in
            guard let self else {
                return
            }
            guard shouldClose else {
                return
            }
            self.confirmedCloseInProgress = true
            sender?.close()
        }
        return false
    }

    private func findEditorViewController(in viewController: NSViewController?) -> EditorViewController? {
        guard let viewController else {
            return nil
        }
        if let editorViewController = viewController as? EditorViewController {
            return editorViewController
        }
        for child in viewController.children {
            if let editorViewController = findEditorViewController(in: child) {
                return editorViewController
            }
        }
        return nil
    }
}
